xray/plot/plot.py

```diff
@@ -257,7 +260,55 @@ def _determine_cmap_params(plot_data, vmin=None, vmax=None, cmap=None,
         else:
             extend = 'neither'
 
-    return vmin, vmax, cmap, extend
+    if levels:
+        cmap, cnorm = _determine_discrete_cmap_params(cmap, levels,
+                                                      vmin, vmax,
+                                                      extend)
+    else:
+        cnorm = None
+
+    return vmin, vmax, cmap, extend, cnorm
+
+
+def _determine_discrete_cmap_params(cmap, levels, vmin, vmax, extend):
+    """
+    Build a descrete colormap and normalization of the data.
+    """
+    import matplotlib as mpl
+    import matplotlib.pyplot as plt
+
+    if extend == 'both':
+        ext_n = 2
+    elif extend in ['min', 'max']:
+        ext_n = 1
+    else:
+        ext_n = 0
+
+    if isinstance(levels, int):
+        cticks = np.linspace(vmin, vmax, num=levels + 1, endpoint=True)
```

Comment:
I wonder if we can deal with the non-colored endpoints issue by simply adding some small epsilon to the last tick-point?